triton/nonblocking_stream.py

```diff
@@ -0,0 +1,174 @@
+# being told to exit.
+LINGER_SHUTDOWN_MSECS = 3000
+
+META_STRUCT_FMT = "!B64p64p"
+
+# We're going to include a version byte in our meta struct for future
+# upgrading.  This is very quickly getting into the kind of bit packing I
+# wanted to avoid for a logging infrastructure, but the performance gain is
+# hard to ignore.
+META_STRUCT_VERSION = 0x4
+
+
+def check_meta_version(meta):
+    value, = struct.unpack(">B", meta[0])
+    if value != META_STRUCT_VERSION:
+        raise ValueError(value)
+
+
+def msgpack_encode_default(obj):
+    """Extra encodings for python types into msgpack
+
+    These are attempted if our normal serialization fails.
+    """
+    if isinstance(obj, decimal.Decimal):
+        return str(obj)
+    if isinstance(obj, datetime.datetime):
+        return time.mktime(obj.utctimetuple())
+    if isinstance(obj, datetime.date):
+        return obj.strftime("%Y-%m-%d")
+
+    raise TypeError("Unknown type: %r" % (obj,))
+
+
+threadLocal = threading.local()
+
+# Context can be shared between threads
+_zmq_context = None
+_connect_str = None
+
+
+def init(host, port):
+    global _zmq_context
+    global _connect_str
+
+    _zmq_context = zmq.Context()
+    _connect_str = "tcp://%s:%d" % (host, port)
+
+
+def _thread_connect():
+    if _zmq_context and not getattr(threadLocal, 'zmq_socket', None):
+        threadLocal.zmq_socket = _zmq_context.socket(zmq.PUSH)
+        threadLocal.zmq_socket.hwm = MAX_QUEUED_MESSAGES
+        threadLocal.zmq_socket.linger = LINGER_SHUTDOWN_MSECS
+
+        threadLocal.zmq_socket.connect(_connect_str)
+
+
+class NonblockingStream(object):
+
+    def __init__(self, name, partition_key):
+        self.name = name
+        if len(self.name) > 64:
+            raise ValueError("Stream Name Too Long")
+        self.partition_key = partition_key
+        if _zmq_context is None:
+            init(*config.get_zmq_config())
+
+    def _partition_key(self, data):
+        return str(data[self.partition_key])
+
+    def _serialize_context(self, data):
+        # Our sending format is made up of two messages. The first has a
+        # quick to unpack set of meta data that our collector is going to
+        # use for routing and stats. This is much faster than having the
+        # collector decode the whole event. We're just going to use python
+        # struct module to make a quick and dirty data structure
+        if len(self._partition_key(data)) > 64:
+            raise ValueError("Partition Key Too Long")
+
+        meta_data = struct.pack(META_STRUCT_FMT, META_STRUCT_VERSION,
+                                self.name,
+                                self._partition_key(data))
+
+        try:
+            message_data = msgpack.packb(data)
+        except TypeError:
+            try:
+                # If we fail to serialize our context, we can try again with an
+                # enhanced packer (it's slower though)
+                message_data = msgpack.packb(data,
```

Comment:
This pattern (of trying, then re-trying with a slower more robust method on failure) is direct from BlueOx. I believe that it is an efficiency optimization; no need to replace the default encoder unless there is actually a time, date, or decimal object. It isn't clear to me how we are handing this kind of scenario in Triton currently; it may be that we are not sending those data types yet and haven't run into that scenario.

I think this pattern is useful here (reasonably common data types); happy to extend this pattern back to the other places in triton where we encode using mspack so we have that uniformity.